### ISSUE TITLE
hp DoFHandler duplicates dofs on edges in 3D.

### ISSUE THREAD

**Issue by mwichro** [dealii/dealii#17861]
In 3D for elements of order 2 and above on  some nodes on edges some degrees of freedom are duplicated. The attached program classifies the mesh, assigns fe index to each element and outputs locations of dofs for each component (6 files).
(extension changed to txt because GitHub does not accept .cc):
[test_hpdh.txt](https://github.com/user-attachments/files/17720257/test_hpdh.txt)
After running the program do
grep '"[0-9]*, [0-9]*"' dof_locations-1.gnuplot
to list the nodes with duplicated DoFs:
-0.75 -0.75 -1.29271 "92, 608"

-0.75 -0.75 -0.957295 "93, 609"
(the list goes on, there are 144 lines)

The problem does not occur for degree 1 (most likely because it is connected to edges), nor in 2D.

**Comment by kronbichler** [comment]
As far as I remember, is the expected behavior, for hp adaptivity the dofs on the lines are not combined. The strategy is generally to combine the DoFs via the constraints, where the 'duplicated' degrees of freedom will get identity constrained to each other. I agree it is a bit unfortunate to have this behavior in your case where there is not actually a different polynomial degree that would motivate the un-merged degrees of freedom. But on the other hand, most algorithms do not care too much. Do you have an algorithm that is severely affected? In that case, I would probably try to match identity constraints directly.

**Comment by mwichro** [comment]
> Do you have an algorithm that is severely affected? In that case, I would probably try to match identity constraints directly.
Yes. I am came across that when I was running high-order cutFEM, where the ghost penalty

**Comment by bangerth** [comment]
The issue with unifying DoFs on edges in 3d is describe in the 2007 paper by myself and Oliver Kayser-Herold, which you can find on my website. I don't recall the details, but they're in the paper.

If I understand you right, then you have an edge at which cells with different `active_fe_index` meet. So you have different elements, but one vector component of these different elements is the same. You expect that the corresponding DoFs are identified. I think that in practice, it is going to be very difficult to figure out this specific case. Generally, `DoFHandler` only deals with `FiniteElement` information, without looking into whether it may be an `FESystem` that could be taken apart in some specific ways. One *could* of course think about whether it is worth treating your specific case differently from the generic code, but I suspect that the actual savings would be fairly minimal -- the way we deal with this right now is to simply enumerate two DoFs, and then add a constraints to the `AffineConstraints` object saying that the two are identical.
